### DISCUSSION TITLE
Backend Convention

### DISCUSSION THREAD

**Discussion by ghrltjdtprbs** [10aeat/10aeat_Backend#3] [공지사항]
## Code Convention

---

- [[IntelliJ Java Google Style](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml)](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml)을 준수합니다.
    
    [[Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)](https://google.github.io/styleguide/javaguide.html)
    
    - `Preferences > Editor > Code Style > Java > Scheme Import > Google Style IDE`
    - `Tab and Indents`에서 Indent 및 ***Tab Size를 4로 변경***합니다.
    - 변경이 완료되었으면 단축키를 이용해 코드를 맞출 수 있습니다.
    
            → 윈도우 : `***Ctrl + Alt + L***`
    
            → 맥       : `***Cmd + Option+ L***`
    

## Package

**java : 17**

**springboot : 3.1.5**

```
com.example
├── domain
│   ├── user
│   ├── accommodation
│   ├── review
│   ├── reservation
│   ├── accommodationLikes
│   ├── basket
│   ...
└── global
    ├── auth
    ├── common
    ├── config
    ├── error
    ├── infra
    ├── util
    └── entity
```

## ariable Convention

- 기본적인 도메인 변수는 ERD에서 작성한 변수명을 따른다.

## Method Convention

- boolean 타입 반환 메서드 : `boolean isXXX( )`

```java
boolean isOutOfStock() // 재고가 다 소진되었는지
boolean isNull()       // Null 값 인지?
```

## Test Convention

```java
@DisplayName("signUp()은 ")
@Nested
class Context_SignUp {

    @DisplayName("회원가입을 할 수 있다.")
    @Test
    void _willSuccess() throws Exception {
        // given
        SignUpRequest requestDto = new SignUpRequest("user11", "password",
                "password", "테스트유저1");

        // when then
        mockMvc.perform(
                    post("/api/v1/signUp")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(requestDto)))
                .andDo(print())
                .andExpect(status().isCreated());
    }

    @DisplayName("username 은 ")
    @Nested
    class Element_username {
        @DisplayName("null이면 실패한다.")
        @Test
        void null_willFail() throws Exception {
            // given
            SignUpRequest requestDto = new SignUpRequest(null, "password",
                    "password", "테스트유저1");

            // when then
            mockMvc.perform(
                        post("/api/v1/signUp")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(requestDto)))
                    .andDo(print())
                    .andExpect(status().isBadRequest())
                    .andExpect(jsonPath("$.message").value("아이디를 입력해주세요."));
            }
        }
}
```

### 공통 사항

- 다른사람들이 어떤 테스트가 실패했는지 한눈에 알아 볼 수 있게 `@DisplayName(””)` 설정하기.
- 테스트 대상 메서드에서 분기가 존재한다면 `@NestedClass` 사용하기
    - Class Name Prefix : `Context_`
    - Dto Validation 검증 Prefix : `Element_`

# Git

---

### **앞에 태그 붙여주세요!**

<img width="188" alt="Untitled (2)" src="https://github.com/10aeat/10aeat_Backend/assets/105612931/fa139a46-620f-42ef-855f-4cb2d9473529">

- 커밋은 소문자

- **[feat]** : 새로운 기능을 추가하는 경우
- **[fix]** : 버그를 고친경우
- **[docs]** : 문서를 수정한 경우
- **[style]** : 코드 포맷 변경, 세미콜론 누락, 코드 수정이 없는경우
- **[refactor]** : 코드 리펙토링
- **[test]** : 테스트 코드. 리펙토링 테스트 코드를 추가했을 때
- **[rename]** : 파일명(or 폴더명) 을 수정한 경우
- **[remove]** : 코드(파일) 의 삭제가 있을 때. "Clean", "Eliminate" 를 사용하기도 함

예시)

```
[feat] 인증코드 유효시간 만료 exception 추가]

문제 사항 :
 - 인증번호가 불일치하는 예외와 인증번호가 만료된 예외를 분리해야 할 필요성

변경 사항 :
 - VerificationCodeExpiredException 클래스 추가
 - ErrorCode 클래스에 EMAIL_VERIFICATION_CODE_EXPIRED 추가
 - redis에 사용자 이메일이 있으나, 인증번호가 일치하지 않을때는 기존 BAD_REQUEST를, 인증번호가 만료되어 redis에 해당 이메일이 없을 경우 GONE 예외를 발생시킵니다.
```

### **GitFlow-Plugin 사용법**

**기능개발**

1. 현재 브랜치가 `develop` 인 것을 확인한 후 `Start Feature` 선택 (사진1)
2. branch 네이밍 설정하기 (사진2)
3. 신나게 기능개발
4. (필수) **커밋 꼭 하기**
5. `Finish Feature` 선택
6. GitFlow-Plugin이 해당 로직을 수행함
- **git switch(checkout) develop**
- **git merge feature-branch**
- **git branch -d feature-branch**

7 . 현재 브랜치가 develop에 있는 것을 본 후에 github로 push!

### 주의사항

- 항상 **GitFlow 사용 전 Branch 확인**하기!
- `Finish Feature` 누르기 전 **Commit이 완료되었는지 체크**하기 (아니면 브랜치가 삭제되버려요!)

**feature/이슈번호**

### PR 이름에 **커밋과 동일하게 앞에 태그 붙여주세요!**

- PR은 첫 글자 대문자

<img width="847" alt="스크린샷 2024-04-18 15 01 43 (1)" src="https://github.com/10aeat/10aeat_Backend/assets/105612931/18aafa95-556b-4ccf-bc50-6b7835a13f35">

팀원 모두의 리뷰를 받아야 merge할 수 있도록 브랜치 규칙 설정

 Pull requests → 리뷰할 pr선택 → 상단의 Add your review →    Review changes → 리뷰 작성 → Approve 선택 → Submit review

<img width="185" alt="Untitled (3)" src="https://github.com/10aeat/10aeat_Backend/assets/105612931/cc98497c-6d78-4dbf-ab19-26c556cbe4ae">

- 옆에 체크 표시가 떠야 리뷰 완료
- 취소는 체크 표시 옆  🔄️ 선택

**Comment by yerim08** [comment]
깃 컨벤션 정리 넘 잘해두셨고 일관성 있게 쓰면 좋을 것 같아서 프론트도 같은 컨벤션 사용하고 싶은데 괜찮을까용 ??

**Reply by ghrltjdtprbs** [reply]
고럼요😄
